src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,405 @@
+	 *
+	 * @param string $plugin The plugin to check.
+	 *
+	 * @throws \Exception If errors are present.
+	 *
+	 * @return void
+	 */
+	private function check_plugin_for_errors( $plugin ) {
+		global $wp_filesystem;
+
+		if ( file_exists( ABSPATH . '.maintenance' ) ) {
+			$wp_filesystem->delete( ABSPATH . '.maintenance' );
+		}
+
+		$nonce    = wp_create_nonce( 'plugin-activation-error_' . $plugin );
+		$response = wp_remote_get(
+			add_query_arg(
+				array(
+					'action'   => 'error_scrape',
+					'plugin'   => $plugin,
+					'_wpnonce' => $nonce,
+				),
+				admin_url( 'plugins.php' )
+			),
+			array( 'timeout' => 60 )
+		);
+
+		if ( is_wp_error( $response ) ) {
+			// If it isn't possible to run the check, assume an error.
+			throw new Exception( $response->get_error_message() );
+		}
+
+		$code           = wp_remote_retrieve_response_code( $response );
+		$body           = wp_remote_retrieve_body( $response );
+		$this->no_error = 200 === $code;
+
+		if ( str_contains( $body, 'wp-die-message' ) || 200 !== $code ) {
+			$error = new WP_Error(
+				'new_version_error',
+				sprintf(
+					/* translators: %s: The name of the plugin. */
+					__( 'The new version of %s contains an error' ),
+					get_plugin_data( WP_PLUGIN_DIR . '/' . $plugin )['Name']
+				)
+			);
+			throw new \Exception( $error->get_error_message() );
+		}
+	}
+
+	/**
+	 * Initializes handlers.
+	 */
+	private function initialize_handlers() {
+		// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_set_error_handler
+		set_error_handler( array( $this, 'error_handler' ), ( E_ALL ^ $this->error_types ) );
+		set_exception_handler( array( $this, 'exception_handler' ) );
+	}
+
+	/**
+	 * Handles Errors.
+	 */
+	public function error_handler() {
+		$this->handler_args['handler_error'] = 'Error Caught';
+		$this->handler( $this->non_fatal_errors() );
+	}
+
+	/**
+	 * Handles Exceptions.
+	 */
+	public function exception_handler() {
+		$this->handler_args['handler_error'] = 'Exception Caught';
+		$this->handler( false );
+	}
+
+	/**
+	 * Handles errors by running Rollback.
+	 *
+	 * @param bool $skip If false, assume fatal and process.
+	 *                   Default false.
+	 */
+	private function handler( $skip = false ) {
+		if ( $skip ) {
+			return;
+		}
+		self::$fatals[] = $this->handler_args['hook_extra']['plugin'];
+
+		$this->cron_rollback();
+
+		/**
+		 * This possibly helps to avoid a potential race condition on servers that may start to
+		 * process the next plugin for auto-updating before the handler can pick up an error from
+		 * the previously processed plugin.
+		 */
+		sleep( 2 );
+
+		$this->restart_updates();
+		$this->restart_core_updates();
+		$this->send_update_result_email();
+	}
+
+	/**
+	 * Skip (exit handler() early) for non-fatal errors or non-errors.
+	 *
+	 * @return bool
+	 */
+	private function non_fatal_errors() {
+		$e                = error_get_last();
+		$non_fatal_errors = ( ! empty( $e ) && $this->error_types !== $e['type'] );
```

Comment:
```suggestion
		$last_error       = error_get_last();
		$non_fatal_errors = ( ! empty( $last_error ) && $this->error_types !== $e['type'] );
```
We've been generally trying to avoid single-character variable names for the sake of readability.

Reply:
Also, since `$non_fatal_errors` is `bool`, maybe `$has_non_fatal_errors` is clearer. I thought for a second this held the non-fatal errors.